cpp_guidelines/guidelines.md

```diff
@@ -0,0 +1,496 @@
+
+**Enforced by:**
+  * `.clang-format`: `ContinuationIndentWidth: 4`
+
+**Example:**
+```c++
+auto foo = some_long_initializer{value1,
+    value2, {inner_value_1,
+        inner_value_2}};
+```
+
+### Case labels are not indented.
+
+**Reason**: Case labels do not open a new scope. Indentation helps to
+differentiate between scopes.
+
+**Enforced by:**
+  * `.clang-format`: `IndentCaseLabels: false`
+
+**Example:**
+```c++
+switch(statement)
+{
+case first_case:
+    /* something */
+    break;
+
+case second_case:
+    /* something else */
+    break;
+}
+```
+
+### Case blocks are indented.
+
+**Reason**: Case blocks are part of a scope (opened by `switch`) and should be
+indented accordingly.
+
+**Enforced by:**
+  * `.clang-format`: `IndentCaseBlocks: true`
+
+```c++
+switch(statement)
+{
+case first_case:
+    /* something */
+    break;
+
+case second_case:
+    /* something else */
+    break;
+}
+```
+
+### `extern "C"` blocks are indented.
+
+**Reason:** `extern "C"` opens a new scope (with special semantics). It
+should be indented accordingly.
+
+**Enforced by:**
+  * `.clang-format`: `IndentExternBlock: AfterExternBlock`
+
+### `goto` labels are not indented.
+
+***DO NOT USE `goto`!***
+
+**Reason:** goto labels do not open scopes. There is no reason to add an
+extra level of indentation to the following lines.
+
+**Enforced by:**
+  * `.clang-format`: `IndentGotoLabels: false`
+
+### Nested preprocessor directives are indented without the hash.
+
+**Reason:** This helps to easily identify lines that target the preprocessor.
+
+**Enforced by:**
+  * `.clang-format`: `IndentPPDirectives: AfterHash`
+
+**Example:**
+```c++
+#if defined(something)
+#   if defined(something_else)
+#       define FOO BAR
+#   else
+#       define FOO BAZ
+#   endif
+#endif
+```
+
+### Inheritance lists and constructor initializers are indented.
+**Reason:** They are part of the class / constructor scope and thus are
+indented.
+
+**Enforced by:**
+  * `.clang-format`: `ConstructorInitializerIndentWidth: 4`
+
+### Class access modifiers are not indented.
+
+**Reason:** Indention marks a scope. Class access modifiers do not open
+separate scopes. This also follows the style of the C++ standard and is the
+default setting of many IDEs.
+
+**Enforced by:**
+  * `.clang-format`: `AccessModifierOffset: -4`
+
+**Example:**
+```c++
+class foo
+{
+public:
+    /* The interface */
+private:
+    /* The secrets. */
+};
+```
+
+### Namespace members and nested namespaces are indented.
+
+**Reason:** This follows the overall code style.
+
+**Enforced by:**
+  * `.clang-format`: `NamespaceIndentation: All`
+
+## Code style
+
+### Align pointers and references to their type.
+
+**Reason:** Pointers and references are part of the type and should be
+part of the declaration as well.
+
+**Enforced by:**
+  * `.clang-format`: `DerivePointerAlignment: false`
+  * `.clang-format`: `PointerAlignment: left`
+
+**Example:**
+```c++
+int *bad;
+int * worse;
+int* good;
```

Comment:
Just noticed this one: This evaluation is of which position is good and which bad actually goes counter to semantics because the `*` belong goes with the identifier, not the type:
```c++
int* ptr, notPtr;
```
Aligning left consistently would lead to this:
```c++
int* ptrA,* ptrB;
```
which looks very strange.

Personally, I also lean towards
```c++
int* pLeft;
```
for astethical reasons, I guess. But this makes be inconsistent with the alignment. Maybe
```c++
int *pRight;
```
is better?

**VOTE**
current: left | right |
--------|-----------|
 0      |  0.5  |

Reply:
current: left | right |
---------------|----------|
1 | 0.5 |

I don't really see the use case for multiple declarations in a single line.

Reply:
current: left | center | right
-- | -- | --
1 | 1 | 0.5

since no parenthesis are supported, the only easily readable way to declare multiple pointers in one line is making a new type first.
If we try to maximize consistency, the stars should also be separated with a space

```c++
using IntPtr = int *;
IntPtr ptrA, ptrB;
```

Reply:
current: left | center | right
-- | -- | --
3 | 1 | 0.5

Reply:
current: left | center | right
-- | -- | --
2 | 2 | 0.5

As a side note, I am not sure why the current version is left. In PIConGPU we definitely use the center formatting as of now.

Reply:
> As a side note, I am not sure why the current version is left.

Because I based the initial clang-format on the MSVC style and (in part) my own preferences, not the current (varying) styles of our projects.

Reply:
Voting closed, the winner is `left`.